### PR TITLE
[CARBONDATA-2524] Support create carbonReader with default projection

### PR DESCRIPTION
**Pull request by xubo245**
1.Support  create carbonReader with default projection, like:
`  CarbonReader reader = CarbonReader.builder(path, "_temp").build();
`
2.Add projectAllColumns method, like:
`  CarbonReader reader = CarbonReader.builder(path, "_temp")
        . projectAllColumns().build();
`
Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [x] Any interfaces changed?
 NA
 - [x] Any backward compatibility impacted?
 NO
 - [x] Document update required?
NA
 - [x] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       Yes, add some test case.
 - [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
No

### CODE REVIEW

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5081/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6082/

**Comment by xubo245** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4922/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5083/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6109/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4947/

**Comment by xubo245** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6111/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4950/

File: store/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonReaderTest.java [hunk 1]
```diff
@@ -409,4 +409,76 @@ public boolean accept(File pathname) {
...
+
+    Field[] fields = new Field[2];
+    fields[0] = new Field("name", DataTypes.STRING);
+    fields[1] = new Field("age", DataTypes.INT);
+
+    TestUtil.writeFilesAndVerify(new Schema(fields), path, true);
+
+    CarbonReader reader = CarbonReader
+        .builder(path, "_temp")
+        .projectAllColumns()
```
Comment (sounakr) [inline store/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonReaderTest.java]: Please add a test case pass null string to projection API. for e.g.
.projection(new String[]{})

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReaderBuilder.java [hunk 2]
```diff
@@ -152,9 +179,10 @@ public CarbonReaderBuilder setEndPoint(String value) {
     if (filterExpression != null) {
       format.setFilterPredicates(job.getConfiguration(), filterExpression);
     }
-    if (projectionColumns != null) {
-      format.setColumnProjection(job.getConfiguration(), new CarbonProjection(projectionColumns));
+    if (projectionColumns == null) {
```
Comment (sounakr) [inline store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReaderBuilder.java]: Rather than checking projectionColumns better to set a member variable like projectAllColumns when projectAllColumns() API is being used. Only incase projectAllColumns is show all columns. projectionColumns can be null in many other cases also.

File: store/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonReaderTest.java (hunk 1 above)
Comment (xubo245) [reply store/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonReaderTest.java]: ok, done. It return null object.

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReaderBuilder.java (hunk 2 above)
Comment (xubo245) [reply store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReaderBuilder.java]: If changed like you said, it will not support this build method :

    CarbonReader reader = CarbonReader
        .builder(path, "_temp")
        .build();

Test case:

  @Test
  public void testReadFilesWithDefaultProjection() throws IOException, InterruptedException {
    String path = "./testWriteFiles";
    FileUtils.deleteDirectory(new File(path));

    Field[] fields = new Field[2];
    fields[0] = new Field("name", DataTypes.STRING);
    fields[1] = new Field("age", DataTypes.INT);

    TestUtil.writeFilesAndVerify(new Schema(fields), path, true);

    CarbonReader reader = CarbonReader
        .builder(path, "_temp")
        .build();

    // expected output after sorting
    String[] name = new String[100];
    int[] age = new int[100];
    for (int i = 0; i < 100; i++) {
      name[i] = "robot" + (i / 10);
      age[i] = (i % 10) * 10 + i / 10;
    }
    // Default sort column is applied for dimensions. So, need  to validate accordingly

    int i = 0;
    while (reader.hasNext()) {
      Object[] row = (Object[]) reader.readNextRow();
      // Default sort column is applied for dimensions. So, need  to validate accordingly
      Assert.assertEquals(name[i], row[0]);
      Assert.assertEquals(age[i], row[1]);
      i++;
    }
    Assert.assertEquals(i, 100);
  }

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5107/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6114/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4952/

**Comment by xubo245** [comment]
@sounakr Done, CI pass.

**Comment by sounakr** [comment]
LGTM

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReaderBuilder.java [hunk 3]
```diff
@@ -61,6 +64,33 @@ public CarbonReaderBuilder projection(String[] projectionColumnNames) {
     return this;
   }
 
+  /**
+   * All Columns need to be Projected
+   *
+   * @return CarbonReaderBuilder object
+   * @throws IOException
+   */
+  public CarbonReaderBuilder projectAllColumns() throws IOException {
```
Comment (kunal642) [inline store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReaderBuilder.java]: can you add a description for this method in reader document

Comment (xubo245) [reply store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReaderBuilder.java]: ok, done

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6130/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4968/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6135/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4971/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5119/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5120/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5123/
